src/topology/algebra/module/locally_convex.lean

```diff
@@ -0,0 +1,138 @@
+  [topological_space E] : Prop :=
+(convex_basis : ∀ x : E, (𝓝 x).has_basis (λ (s : set E), s ∈ 𝓝 x ∧ convex 𝕂 s) id)
+
+variables (𝕂 E : Type*) [ordered_semiring 𝕂] [add_comm_monoid E] [module 𝕂 E] [topological_space E]
+
+lemma locally_convex_space_iff :
+  locally_convex_space 𝕂 E ↔
+  ∀ x : E, (𝓝 x).has_basis (λ (s : set E), s ∈ 𝓝 x ∧ convex 𝕂 s) id :=
+⟨@locally_convex_space.convex_basis _ _ _ _ _ _, locally_convex_space.mk⟩
+
+lemma locally_convex_of_bases {ι : Type*} (b : E → ι → set E) (p : ι → Prop)
+  (hbasis : ∀ x : E, (𝓝 x).has_basis p (b x)) (hconvex : ∀ x i, p i → convex 𝕂 (b x i)) :
+  locally_convex_space 𝕂 E :=
+⟨λ x, (hbasis x).to_has_basis
+  (λ i hi, ⟨b x i, ⟨⟨(hbasis x).mem_of_mem hi, hconvex x i hi⟩, le_refl (b x i)⟩⟩)
+  (λ s hs, ⟨(hbasis x).index s hs.1,
+    ⟨(hbasis x).property_index hs.1, (hbasis x).set_index_subset hs.1⟩⟩)⟩
+
+lemma locally_convex_space.convex_basis_zero [locally_convex_space 𝕂 E] :
+  (𝓝 0 : filter E).has_basis (λ s, s ∈ (𝓝 0 : filter E) ∧ convex 𝕂 s) id :=
+locally_convex_space.convex_basis 0
+
+lemma locally_convex_iff_exists_convex_subset :
+  locally_convex_space 𝕂 E ↔ ∀ x : E, ∀ U ∈ 𝓝 x, ∃ S ∈ 𝓝 x, convex 𝕂 S ∧ S ⊆ U :=
+begin
+  rw locally_convex_space_iff,
+  refine forall_congr (λ x, _),
+  simp_rw [has_basis_iff, exists_prop, and_assoc],
+  exact ⟨λ H U, (H U).mp, λ H U, ⟨H U, λ ⟨S, hS, _, hS'⟩, mem_of_superset hS hS'⟩⟩
+end
+
+end semimodule
+
+section module
+
+variables (𝕂 E : Type*) [ordered_semiring 𝕂] [add_comm_group E] [module 𝕂 E] [topological_space E]
+  [topological_add_group E]
+
+lemma locally_convex_of_basis_zero {ι : Type*} (b : ι → set E) (p : ι → Prop)
+  (hbasis : (𝓝 0).has_basis p b) (hconvex : ∀ i, p i → convex 𝕂 (b i)) :
+  locally_convex_space 𝕂 E :=
+begin
+  refine locally_convex_of_bases 𝕂 E (λ (x : E) (i : ι), ((+) x) '' b i) p (λ x, _)
+    (λ x i hi, (hconvex i hi).translate x),
+  rw ← map_add_left_nhds_zero,
+  exact hbasis.map _
+end
+
+lemma locally_convex_space_iff_zero :
+  locally_convex_space 𝕂 E ↔
+  (𝓝 0 : filter E).has_basis (λ (s : set E), s ∈ (𝓝 0 : filter E) ∧ convex 𝕂 s) id :=
+⟨λ h, @locally_convex_space.convex_basis _ _ _ _ _ _ h 0,
+ λ h, locally_convex_of_basis_zero 𝕂 E _ _ h (λ s, and.right)⟩
+
+lemma locally_convex_iff_exists_convex_subset_zero :
+  locally_convex_space 𝕂 E ↔
+  ∀ U ∈ (𝓝 0 : filter E), ∃ S ∈ (𝓝 0 : filter E), convex 𝕂 S ∧ S ⊆ U :=
+begin
+  rw locally_convex_space_iff_zero,
+  simp_rw [has_basis_iff, exists_prop, and_assoc],
+  exact ⟨λ H U, (H U).mp, λ H U, ⟨H U, λ ⟨S, hS, _, hS'⟩, mem_of_superset hS hS'⟩⟩
+end
+
+end module
+
+-- TODO : generalize the following section to normed and ordered rings/fields, once we have the
+-- corresponding typeclasses
+
+namespace seminorm
+
+variables {ι 𝕂 E : Type*} [add_comm_group E] [module ℝ E] [nonempty ι] [topological_space E]
+  [topological_add_group E]
+
+lemma locally_convex_space (p : ι → seminorm ℝ E) [with_seminorms p] :
```

Comment:
```suggestion
instance _root_.with_seminorms.to_locally_convex_space (p : ι → seminorm ℝ E) [with_seminorms p] :
```

Reply:
I made that a lemma because the linter complains that it is a dangerous instance because the `ι` can't be inferred. Should I nolint it ?

Reply:
About the name, `with_seminorms` is in the `seminorm` namespace, so the full name should be `seminorm.with_seminorms.to_locally_convex_space`, right ?

Reply:
`with_seminorms` *shouldn't be in the `seminorm` namespace. I missed that in the review 😅 

Calling @Vierkantor for the instance trickery.